src/core/harpoon.ml

```diff
@@ -304,14 +304,130 @@ module Tactic (T : TacticContext) = struct
       ; solution = None
       }
     in
-    T.remove_current_subgoal ();
-    T.add_subgoal new_state;
+    tctx.remove_current_subgoal ();
+    tctx.add_subgoal new_state;
     prepend_commands
       [ Comp.IH (m, name) ]
       (Comp.incomplete_proof new_state)
     |> solve' g
 end
 
+module Automation = struct
+  open Comp
+
+  type t = unit proof_state -> Tactic.tactic_context -> bool
+
+  let auto_intros : t =
+    fun g tctx ->
+    let (t, _) = g.goal in
+    dprintf
+      (fun p ->
+        p.fmt
+          "[auto_intros]: invoked on %a"
+          (P.fmt_ppr_cmp_typ g.context.cD Pretty.std_lvl) t
+      );
+    match t with
```

Comment:
Perhaps instead of matching on the type to decide whether to do intros or not, you can adjust `Tactic.intros` so it returns a bool; then `auto_intros` just becomes `intros`. Note that `intros` is should always be safe to call: it should just do nothing if the goal doesn't have a function type.